impl/src/main/java/org/ehcache/impl/persistence/DefaultDiskResourceService.java

```diff
@@ -0,0 +1,274 @@
+   */
+  @Override
+  public PersistenceSpaceIdentifier<DiskResourceService> getPersistenceSpaceIdentifier(String name, CacheConfiguration<?, ?> config) throws CachePersistenceException {
+    if (persistenceService == null) {
+      return null;
+    }
+    boolean persistent = config.getResourcePools().getPoolForResource(ResourceType.Core.DISK).isPersistent();
+    while (true) {
+      PersistenceSpace persistenceSpace = knownPersistenceSpaces.get(name);
+      if (persistenceSpace != null) {
+        return persistenceSpace.identifier;
+      }
+      PersistenceSpace newSpace = createSpace(name, persistent);
+      if (newSpace != null) {
+        return newSpace.identifier;
+      }
+    }
+  }
+
+  @Override
+  public void releasePersistenceSpaceIdentifier(PersistenceSpaceIdentifier<?> identifier) throws CachePersistenceException {
+    String name = null;
+    for (Map.Entry<String, PersistenceSpace> entry : knownPersistenceSpaces.entrySet()) {
+      if (entry.getValue().identifier.equals(identifier)) {
+        name = entry.getKey();
+      }
+    }
+    if (name == null) {
+      throw new CachePersistenceException("Unknown space " + identifier);
+    }
+    PersistenceSpace persistenceSpace = knownPersistenceSpaces.remove(name);
+    if (persistenceSpace != null) {
+      for (FileBasedStateRepository stateRepository : persistenceSpace.stateRepositories.values()) {
+        try {
+          stateRepository.close();
+        } catch (IOException e) {
+          LOGGER.warn("StateRepository close failed - destroying persistence space {} to prevent corruption", identifier, e);
+          persistenceService.destroySafeSpace(((DefaultPersistenceSpaceIdentifier)identifier).persistentSpaceId, true);
+        }
+      }
+    }
+  }
+
+  private PersistenceSpace createSpace(String name, boolean persistent) throws CachePersistenceException {
+    DefaultPersistenceSpaceIdentifier persistenceSpaceIdentifier =
+        new DefaultPersistenceSpaceIdentifier(persistenceService.createSafeSpaceIdentifier(PERSISTENCE_SPACE_OWNER, name));
+    PersistenceSpace persistenceSpace = new PersistenceSpace(persistenceSpaceIdentifier);
+    if (knownPersistenceSpaces.putIfAbsent(name, persistenceSpace) == null) {
+      boolean created = false;
+      try {
+        if (!persistent) {
+          persistenceService.destroySafeSpace(persistenceSpaceIdentifier.persistentSpaceId, true);
+        }
+        persistenceService.createSafeSpace(persistenceSpaceIdentifier.persistentSpaceId);
+        created = true;
+      } finally {
+        if (!created) {
+          // this happens only if an exception is thrown..clean up for any throwable..
+          knownPersistenceSpaces.remove(name, persistenceSpace);
+        }
+      }
+      return persistenceSpace;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void destroy(String name) throws CachePersistenceException {
+    if (persistenceService == null) {
```

Comment:
@henri-tremblay May want to give it some thought. Because this is now decomposed in two services, the solution we came up with for 3.1.2 no longer works. However, we should stick with the public behaviour IMO. Just find a different implementation to fix the problem.